src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1434,6 +1433,72 @@
                    (when (and update-latest-report? (not= type "plan"))
                      (update-latest-report! certname report-id producer_timestamp)))))))))
 
+(defn spawn-query-killer
+  "Spawns a thread which will loop and wait for the gc-pid to get blocked
+   waiting on an AccessExclusiveLock. Once this thread detects that GC is
+   blocked it will cancel all running queries from the pdb user against the
+   pdb database which have been granted locks with the exception of queries
+   from the gc-pid or the killer's pid. This should clear the way for GC to
+   grab the lock it's requesting in the main GC thread. This loop repeats
+   until the killer thread is interrupted or receives confirmation from the
+   GC thread via the gc-finished? atom that the GC thread has dropped the
+   partition."
+  [db gc-pid gc-finished?]
+  (let [{:keys [current_database current_user]}
+        (first (jdbc/query-to-vec "select current_database(), current_user;"))
+
+        kill-blocking-qs
+        (fn [current_database current_user gc-pid]
+          (jdbc/do-prepared
+           (str "select pg_cancel_backend(pid)"
+                " from pg_stat_activity"
+                " where (datname = ? AND usename = ?)"
+                "  and pid in (select unnest(pg_blocking_pids(?)))")
+           [current_database current_user gc-pid]))]
+    (Thread.
+     #(with-noisy-failure
+        (try
+          ;; you can't use *db* value from the GC thread because the connection
+          ;; it has may already be blocked attempting to drop the partition.
+          ;; Binding *db* to the passed in db var will cause the killer thread
+          ;; to get a new Hikari connection
+          (binding [jdbc/*db* db]
+            (while (not @gc-finished?)
+              (kill-blocking-qs current_database
```

Comment:
If we did decide we wanted a 5-10 second grace period before attempting to cancel blocking queries we could add a sleep here before the kill query is issued.